### PR TITLE
[linear algebra] Patch Openblas v0.3.10 to fix EXCEPTION_ACCESS_VIOLATION

### LINKED ISSUE #40963: Exception access violation / openblas / Julia 1.6.1
Hi all, 

I've just started programming in Julia to do numerical simulations, and while testing a function using included I got this message. does anyone have a clue about it? I've tried to search on the web, but since I'm a noob with Julia I didn't really get the point of possible solutions, and this seems very win10 related, so I hope that someone can give me explanations about this. thanks a lot in advance!

Exception: EXCEPTION_ACCESS_VIOLATION at 0x1fd608e0 --  at 0x1fd608e0 -- OLATION with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x1fd608e0 --  at 0x1fd608e0 -- OLATION with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x1fd608e0 --  at 0x1fd608e0 -- OLATION\AppData\Local\Programs\Julia-1.6.1\bin\libopenblas64_.DLL (unknown line)
 their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x1fd608e0 --  at 0x1fd608e0 -- OLATION\AppData\Local\Programs\Julia-1.6.1\bin\libopenblas64_.DLL (unknown line)
in expression starting at REPL[10]:1
in expression starting at REPL[10]:1
cal\Programs\Julia-1.6.1\bin\libopenblas64_.DLL (unknown line)
 their entirety). Thanks.
Exception:

### PR DESCRIPTION
**Pull request by inkydragon**
## Issue Description
OpenBLAS, or more precisely CBLAS, will crash when you multiply two large matrices together.
After testing, this issue was found to occur in OpenBLAS v0.3.10, and was fixed in v0.3.12 and later.

### Conditions for problem reproduction
+ Windows system. Cannot reproduce on macOS, Linux, or even WSL.
+ Intel CPU. Cannot reproduce on AMD CPU.
+ Julia 1.6.x. 1.7.0 use OpenBLAS v0.3.13, the problem is fixed

### Test code
1. Use `LinearAlgebra.BLAS`

Note: You can replace `Julia-1.6.x/bin/libopenblas64_.dll` to test other dlls.
```julia
arrSzie = 10000  # No error
arrSzie = 10001  # Has error
arrType = ComplexF64

using LinearAlgebra.BLAS: dotc, dotu

x = ones(arrType, arrSzie);
y = ones(arrType, arrSzie);

dotc(x, y) # crash: EXCEPTION_ACCESS_VIOLATION
dotu(x, y) # crash

dotc(x, y) == dotu(x, y) == 10001.0 + 0.0im
```

2. call the dll directly from `ccall`
```julia

arrSzie = 10001
arrType = ComplexF64
# OpenBLAS inside julia uses the `:libopenblas64_` symbol, 
#   so please **RENAME** the DLL you want to test to `libopenblas.dll`.
blasPath = raw"V:\libopenblas.dll"
blasSym = :libopenblas

using LinearAlgebra.BLAS: BlasInt
using Libdl: dlopen

dlopen(blasPath)
# Print BLAS version
strip(unsafe_string(ccall((:openblas_get_config64_, blasSym), Ptr{UInt8}, () )))

result = Ref{ComplexF64}();
x = ones(arrType, arrSzie);
y = ones(arrType, arrSzie);
ccall((:cblas_zdotc_sub64_, blasSym), Cvoid,
    (BlasInt, Ptr{ComplexF64}, BlasInt, Ptr{ComplexF64}, BlasInt, Ptr{ComplexF64}),
    length(x), x, stride(x, 1), y, stride(x, 1), result)
# crash
result[]

result[] == 10001.0 + 0.0im
```

You can get the compiled product with the patch applied from the AzurePipeline in JuliaPackaging/Yggdrasil#3655.

e.g.: [OpenBLAS.v0.3.10.x86_64-w64-mingw32-libgfortran5](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=13636&view=logs&j=7cb0c809-6dd4-57ed-8e4b-98cf4b64dcac&t=8580f514-383a-5bc3-ff2b-b06a8ac2d20a&l=31454)

## Patch code

With some `git bisect` (https://github.com/JuliaLang/LinearAlgebra.jl/issues/854), I found that commit b205323 seems to fix this issue.

But, I'm not sure if this fully fixes the issue, more testing may be needed.

----

fix JuliaLang/LinearAlgebra.jl#854.

~~Note: Buildbot uses the compiled product of BinaryBuilder.jl, so **the bug is not fixed in the CI output artifacts**.
You can download the compiled product of BinaryBuilder.jl with patch applied in JuliaPackaging/Yggdrasil#3655, replace the `libopenblas64_.dll` that comes with julia, and test it.~~

~~We may need to merge JuliaPackaging/Yggdrasil#3655 to get the repaired `OpenBLAS_jll.jl` first, and then update the OpenBLAS version in julia's dependency.~~

### CODE REVIEW

**Comment by ViralBShah** [comment]
This looks good. @KristofferC Can we merge?

**Comment by DilumAluthge** [comment]
We should probably merge https://github.com/JuliaPackaging/Yggdrasil/pull/3655 first, right? cc: @staticfloat

**Comment by ViralBShah** [comment]
It doesn't matter. This is just fixing the Julia source build. Once the Yggdrasil PR is merged, we can bump deps/Versions.make and I am sure 2 other places that I always forget.

**Comment by vchuravy** [comment]
Once https://github.com/JuliaRegistries/General/pull/45875 lands we should update versions and refresh checksums

**Comment by inkydragon** [comment]
After updating the version number manually, I updated the checksum with `make -f contrib/refresh_checksums.mk openblas` .

Is there anything else I need to do?

File: deps/checksums/openblas [hunk 1]
```diff
@@ -1,94 +1,94 @@
...
+OpenBLAS.v0.3.10+8.x86_64-w64-mingw32-libgfortran3.tar.gz/md5/a680bfa877cfc313a7e8b2c87469d615
+OpenBLAS.v0.3.10+8.x86_64-w64-mingw32-libgfortran3.tar.gz/sha512/2fd93985e36d62e8d583d07d32e1d37848f47b46dc1f2278909bad98a05c6c47a0ac29fb6217fe017bb66f520b0ef0bde49af370d372d1d8dd43a5f104bd5f7f
+OpenBLAS.v0.3.10+8.x86_64-w64-mingw32-libgfortran4.tar.gz/md5/d616bef77c9d1b0f199ebcb8a5ccf908
+OpenBLAS.v0.3.10+8.x86_64-w64-mingw32-libgfortran4.tar.gz/sha512/b12a9310be4a934b86b073ed03cea18e30f01039399162409b1dff1e097d75430cbaa1d7d5cf0972d32907137c54d455e266aeb854ad86f7bbb309b01c0d9dd1
+OpenBLAS.v0.3.10+8.x86_64-w64-mingw32-libgfortran5.tar.gz/md5/733f0803a1c835a94beeec70ef713512
+OpenBLAS.v0.3.10+8.x86_64-w64-mingw32-libgfortran5.tar.gz/sha512/97fcbfd8af8ad281203529e2d31bf5a4892831ebf2805d09b373ce3bd5fb108f33436ea42b08287d283b7b45e4be0f5433409a9bc5d8252941f30cd03851a839
 openblas-63b03efc2af332c88b86d4fd8079d00f4b439adf.tar.gz/md5/3d692acc6927454f620a4c493bdb159d
 openblas-63b03efc2af332c88b86d4fd8079d00f4b439adf.tar.gz/sha512/cf89f6db1b6366833d29a1dc718ea0b8f61d162f70695c33fc94afbaba232605630a7a7cc3d3b9bed7493ec85402b65180ca99c3101de7141d6f2919318f55c1
```
Comment (inkydragon) [inline deps/checksums/openblas:94]: The above two lines of checksums are not updated, is this the expected behavior of the `contrib/refresh_checksums.mk` script?

**Comment by inkydragon** [comment]
Downloaded and tested buildbot/package_win64 , confirmed that the problem has been fixed

```julia
julia> versioninfo()
Julia Version 1.6.3-pre.85
Commit 10de22d171 (2021-10-01 03:32 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-9400F CPU @ 2.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)

julia> arrSzie = 10000;
julia> arrSzie = 10001;  # Has error
julia> arrType = ComplexF64;

julia> using LinearAlgebra.BLAS: dotc, dotu
julia> x = ones(arrType, arrSzie);
julia> y = ones(arrType, arrSzie);

julia> dotc(x, y) # crash: EXCEPTION_ACCESS_VIOLATION
10001.0 + 0.0im

julia> dotu(x, y) # crash
10001.0 + 0.0im

julia> dotc(x, y) == dotu(x, y) == 10001.0 + 0.0im
true
```

**Comment by inkydragon** [comment]
Tests of `SparseArrays/higherorderfns`, `SparseArrays/sparsevector`, `SparseArrays/sparse`, `Dates/arithmetic`, and `cmdlineargs` failed. 
The logs show that there is a communication problem, which does not seem to be related to this pr.

<details>
<summary>Failed test log</summary>

```
SparseArrays/higherorderfns        (9) |         failed at 2021-10-01T01:16:10.649
ProcessExitedException(9)
Stacktrace:
  [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
    @ Base ./task.jl:710
  [2] wait()
    @ Base ./task.jl:770
  [3] wait(c::Base.GenericCondition{ReentrantLock})
    @ Base ./condition.jl:106
  [4] take_buffered(c::Channel{Any})
    @ Base ./channels.jl:389
  [5] take!(c::Channel{Any})
    @ Base ./channels.jl:383
  [6] take!(::Distributed.RemoteValue)
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:599
  [7] remotecall_fetch(::Function, ::Distributed.Worker, ::String, ::Vararg{String, N} where N; kwargs::Base.Iterators.Pairs{Symbol, UInt128, Tuple{Symbol}, NamedTuple{(:seed,), Tuple{UInt128}}})
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:390
  [8] remotecall_fetch(::Function, ::Int64, ::String, ::Vararg{String, N} where N; kwargs::Base.Iterators.Pairs{Symbol, UInt128, Tuple{Symbol}, NamedTuple{(:seed,), Tuple{UInt128}}})
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:421
  [9] macro expansion
    @ /usr/home/julia/buildbot/w2_tester/tester_freebsd64/build/share/julia/test/runtests.jl:217 [inlined]
 [10] (::var"#29#39"{Vector{Task}, var"#print_testworker_errored#35"{ReentrantLock, Int64, Int64}, var"#print_testworker_stats#33"{ReentrantLock, Int64, Int64, Int64, Int64, Int64, Int64}, Vector{Any}, Dict{String, DateTime}})()
    @ Main ./task.jl:411
SparseArrays/sparsevector          (8) |         failed at 2021-10-01T01:16:10.935
ProcessExitedException(8)
Stacktrace:
  [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
    @ Base ./task.jl:710
  [2] wait()
    @ Base ./task.jl:770
  [3] wait(c::Base.GenericCondition{ReentrantLock})
    @ Base ./condition.jl:106
  [4] take_buffered(c::Channel{Any})
    @ Base ./channels.jl:389
  [5] take!(c::Channel{Any})
    @ Base ./channels.jl:383
  [6] take!(::Distributed.RemoteValue)
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:599
  [7] remotecall_fetch(::Function, ::Distributed.Worker, ::String, ::Vararg{String, N} where N; kwargs::Base.Iterators.Pairs{Symbol, UInt128, Tuple{Symbol}, NamedTuple{(:seed,), Tuple{UInt128}}})
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:390
  [8] remotecall_fetch(::Function, ::Int64, ::String, ::Vararg{String, N} where N; kwargs::Base.Iterators.Pairs{Symbol, UInt128, Tuple{Symbol}, NamedTuple{(:seed,), Tuple{UInt128}}})
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:421
  [9] macro expansion
    @ /usr/home/julia/buildbot/w2_tester/tester_freebsd64/build/share/julia/test/runtests.jl:217 [inlined]
 [10] (::var"#29#39"{Vector{Task}, var"#print_testworker_errored#35"{ReentrantLock, Int64, Int64}, var"#print_testworker_stats#33"{ReentrantLock, Int64, Int64, Int64, Int64, Int64, Int64}, Vector{Any}, Dict{String, DateTime}})()
    @ Main ./task.jl:411
Worker 9 terminated.
Worker 5 terminated.
SparseArrays/sparse                (5) |         failed at 2021-10-01T01:16:11.004
ProcessExitedException(5)
Stacktrace:
  [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
    @ Base ./task.jl:710
  [2] wait()
    @ Base ./task.jl:770
  [3] wait(c::Base.GenericCondition{ReentrantLock})
    @ Base ./condition.jl:106
  [4] take_buffered(c::Channel{Any})
    @ Base ./channels.jl:389
  [5] take!(c::Channel{Any})
    @ Base ./channels.jl:383
  [6] take!(::Distributed.RemoteValue)
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:599
  [7] remotecall_fetch(::Function, ::Distributed.Worker, ::String, ::Vararg{String, N} where N; kwargs::Base.Iterators.Pairs{Symbol, UInt128, Tuple{Symbol}, NamedTuple{(:seed,), Tuple{UInt128}}})
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:390
  [8] remotecall_fetch(::Function, ::Int64, ::String, ::Vararg{String, N} where N; kwargs::Base.Iterators.Pairs{Symbol, UInt128, Tuple{Symbol}, NamedTuple{(:seed,), Tuple{UInt128}}})
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:421
  [9] macro expansion
    @ /usr/home/julia/buildbot/w2_tester/tester_freebsd64/build/share/julia/test/runtests.jl:217 [inlined]
 [10] (::var"#29#39"{Vector{Task}, var"#print_testworker_errored#35"{ReentrantLock, Int64, Int64}, var"#print_testworker_stats#33"{ReentrantLock, Int64, Int64, Int64, Int64, Int64, Int64}, Vector{Any}, Dict{String, DateTime}})()
    @ Main ./task.jl:411
Worker 2 terminated.
Dates/arithmetic                   (2) |         failed at 2021-10-01T01:16:11.067
ProcessExitedException(2)
Stacktrace:
  [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
    @ Base ./task.jl:710
  [2] wait()
    @ Base ./task.jl:770
  [3] wait(c::Base.GenericCondition{ReentrantLock})
    @ Base ./condition.jl:106
  [4] take_buffered(c::Channel{Any})
    @ Base ./channels.jl:389
  [5] take!(c::Channel{Any})
    @ Base ./channels.jl:383
  [6] take!(::Distributed.RemoteValue)
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:599
  [7] remotecall_fetch(::Function, ::Distributed.Worker, ::String, ::Vararg{String, N} where N; kwargs::Base.Iterators.Pairs{Symbol, UInt128, Tuple{Symbol}, NamedTuple{(:seed,), Tuple{UInt128}}})
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:390
  [8] remotecall_fetch(::Function, ::Int64, ::String, ::Vararg{String, N} where N; kwargs::Base.Iterators.Pairs{Symbol, UInt128, Tuple{Symbol}, NamedTuple{(:seed,), Tuple{UInt128}}})
    @ Distributed /usr/home/julia/buildbot/w1_builder/package_freebsd64/build/usr/share/julia/stdlib/v1.6/Distributed/src/remotecall.jl:421
  [9] macro expansion
    @ /usr/home/julia/buildbot/w2_tester/tester_freebsd64/build/share/julia/test/runtests.jl:217 [inlined]
 [10] (::var"#29#39"{Vector{Task}, var"#print_testworker_errored#35"{ReentrantLock, Int64, Int64}, var"#print_testworker_stats#33"{ReentrantLock, Int64, Int64, Int64, Int64, Int64, Int64}, Vector{Any}, Dict{String, DateTime}})()
    @ Main ./task.jl:411
Dates/conversions                 (10) |        started at 2021-10-01T01:16:24.837
Dates/conversions                 (10) |     4.47 |   0.04 |  0.9 |      94.32 |   324.80
LibGit2/online                    (10) |        started at 2021-10-01T01:16:30.385
download                          (11) |        started at 2021-10-01T01:16:33.966
cmdlineargs                        (3) |         failed at 2021-10-01T01:16:37.147
Test Failed at /usr/home/julia/buildbot/w2_tester/tester_freebsd64/build/share/julia/test/cmdlineargs.jl:651
  Expression: occursin(r"\.jl:(\d+)", bt)
   Evaluated: occursin(r"\.jl:(\d+)", "")
```
</details>

File: deps/checksums/openblas (hunk 1 above)
Comment (vchuravy) [reply deps/checksums/openblas:94]: Yes those are the sources

**Comment by vchuravy** [comment]
Thanks!

**Comment by vchuravy** [comment]
(Missed that this was against 1.6)

**Pull request merged**